pkg/resource/deploy/step_generator.go

```diff
@@ -335,8 +378,16 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, res
 		// invalid (they got deleted) so don't consider them. Similarly, if the old resource was External,
 		// don't consider those inputs since Pulumi does not own them. Finally, if the resource has been
 		// targeted for replacement, ignore its old state.
-		if recreating || wasExternal || sg.isTargetedReplace(urn) {
-			inputs, failures, err = prov.Check(urn, nil, goal.Properties, allowUnknowns)
+		if recreating || wasExternal || sg.isTargetedReplace(urn) || !hasOld {
```

Comment:
The `!hasOld` should be safe here, in the case where hasOld == false, oldChecked will be nil and goal.properties == inputs.